File: leetcode/14-longest-common-prefix/main.go
package main

import "fmt"

/*
	questions to ask:
	- empty array?
	- empty string?
*/

/*
	1st approach:
	- append charactors
	Time		O(n*m)
	Space		O(m)
	4ms beats 12.9%
	21jan2019
*/
func longestCommonPrefix1(strs []string) string {
	if len(strs) == 0 {
		return ""
	}
	res := strs[0]
	for i := 1; i < len(strs); i++ {
		temp := ""
		for j := 0; j < len(strs[i]); j++ {
			if j < len(res) && strs[i][j] == res[j] {
				temp += string(res[j])
			} else {
				break
			}
		}
		res = temp
	}
	return res
}

/*
	2nd approach:
	- instead of appending charactors so many times, slice the end
	Time		O(n*m)
	Space		O(m)
	0ms beats 100%
	21jan2019
*/
func longestCommonPrefix(strs []string) string {
	if len(strs) == 0 {
		return ""
	}
	res := strs[0]
	for i := 1; i < len(strs); i++ {
		for j := 0; j < len(strs[i]); j++ {
			if j < len(res) && strs[i][j] != res[j] {
				res = res[:j]
				break
			}
		}
		if len(res) > len(strs[i]) {
			res = res[:len(strs[i])]
		}
	}
	return res
}

func main() {
	a := []string{
		"flower",
		"flow",
		"flight",
	}
	fmt.Println(longestCommonPrefix(a))
	a = []string{
		"dog",
		"racecar",
		"car",
	}
	fmt.Println(longestCommonPrefix(a))
	a = []string{
		"aa",
		"a",
	}
	fmt.Println(longestCommonPrefix(a))
	a = []string{
		"ca",
		"c",
	}
	fmt.Println(longestCommonPrefix(a))
	a = []string{
		"ca",
		"a",
	}
	fmt.Println(longestCommonPrefix(a))
	a = []string{
		"aca",
		"cba",
	}
	fmt.Println(longestCommonPrefix(a))
}

File: leetcode/14-longest-common-prefix/main.py
"""
    2nd approach:
	- instead of appending charactors so many times, slice the end
    
	Time		O(n*m)
	Space		O(m)
	0ms beats 100%
"""


class Solution(object):
    def longestCommonPrefix(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        """
        if len(strs) == 0:
            return ""
        res = strs[0]
        for i in range(1, len(strs)):
            s = strs[i]
            j = 0
            while j < len(res) and j < len(s):
                if s[j] != res[j]:
                    break
                j += 1
            res = s[:j]
        return res

"""
    2nd: greedy
    - get the shortest string
    - iterate that string and for every its own character, to check if there is a mismatch with the others

    Time    O(SN) S:shortest, N:number of strs
    Space   O(1)
"""
class Solution(object):
    def longestCommonPrefix(self, strs):
        if not strs:
            return ""
        shortest = min(strs,key=len)
        for i in range(len(shortest)):
            c = shortest[i]
            for s in strs:
                if s[i] != c:
                    return shortest[:i]
        return shortest